ElectricCar.cpp

```diff
@@ -4,13 +4,48 @@
 ElectricCar::ElectricCar(ElectricEngine* engine)
     : engine_(engine)
 {
-    std::cout << __FUNCTION__ << std::endl;
+    std::cout << "* " << __FUNCTION__ << std::endl;
 }
 
-ElectricCar::~ElectricCar()       { std::cout << __FUNCTION__ << std::endl; }
-void ElectricCar::turnLeft()      { std::cout << __FUNCTION__ << std::endl; }
-void ElectricCar::turnRight()     { std::cout << __FUNCTION__ << std::endl; }
-void ElectricCar::brake()         { std::cout << __FUNCTION__ << std::endl; }
-void ElectricCar::accelerate(int) { std::cout << __FUNCTION__ << std::endl; }
-void ElectricCar::charge()        { std::cout << __FUNCTION__ << std::endl; }
+ElectricCar::~ElectricCar() {
+    std::cout << "* " << __FUNCTION__ << std::endl;
+    delete engine_;
+}
+
+void ElectricCar::charge() { 
+    std::cout << __FUNCTION__;
+    if (chargeLevel_ < MAX_CAPACITY) {
+        chargeLevel_ = MAX_CAPACITY;
+        std::cout << " : recharged" << std::endl;
+    } else {
+        std::cerr << " : battery already full" << std::endl;
+    }
+}
+
+size_t ElectricCar::getChargeLevel() {
+    return chargeLevel_;
+}
 
+void ElectricCar::setEngine(ElectricEngine* engine) {
+    std::cout << __FUNCTION__;
+    if (!getSpeed()) {
+        std::cout << " : engine changed" << std::endl;
+        delete engine_;
+        engine_ = engine;
+        std::cout << "@ electric power: " << engine_->getPower() << std::endl;
+    } else {
+        std::cout << " : not possible unless stopped" << std::endl;
+        delete engine;
+    }
+}
+void ElectricCar::setEngine(PetrolEngine*) {}
+void ElectricCar::setEngine(PetrolEngine*, ElectricEngine*) {}
+
+void ElectricCar::restore() {
+    charge();
+}
+
+void ElectricCar::changeGear(int gear) {
+    std::cout << __FUNCTION__ << " : no transmission" << std::endl;
+    (void)gear;
+}
```

Comment:
Not this way :) In modern C++ lesson there was an attribute [[maybe_unused]] mentioned. Use it instead.
```suggestion
void ElectricCar::changeGear([[maybe_unused]] int gear) {
    std::cout << __FUNCTION__ << " : no transmission" << std::endl;
}
```

Reply:
It was used to get rid of compiler warnings as the variable had not been used (aka placeholder). When I wrote the code there wasn't such a lesson (or I missed that mention). But good to know and applied.

Reply:
I have to watch that lesson (maybe again). Probably in 'Nowoczesny C++ #2' - I haven't watched it yet (I've got Java classes as well and need to share the time accordingly).